src/Features/Core/Portable/FindUsages/SourceReferenceItem.cs

```diff
@@ -34,49 +36,54 @@ internal sealed class SourceReferenceItem
         /// </summary>
         public bool IsWrittenTo { get; }
 
+        public ImmutableArray<CustomColumnInfo> CustomColumns { get; }
+
         /// <summary>
         /// Additional information about the reference.
         /// Each entry represents a key-values pair of data. For example, consider the below entry:
         ///     { "ValueUsageInfo" } = { "Read", "Write" }
         /// This entry indicates that the reference has additional value usage information which indicate
         /// it is a read/write reference, such as say 'a++'.
         /// </summary>
-        public ReferenceInfoMap ReferenceInfo { get; }
+        public ReferenceUsageInfoMap ReferenceUsageInfo { get; }
 
         [Obsolete]
         public SourceReferenceItem(DefinitionItem definition, DocumentSpan sourceSpan, bool isWrittenTo)
         {
             Definition = definition;
             SourceSpan = sourceSpan;
             IsWrittenTo = isWrittenTo;
-            ReferenceInfo = ReferenceInfoMap.Empty;
+            ReferenceUsageInfo = ReferenceUsageInfoMap.Empty;
         }
 
-        public SourceReferenceItem(DefinitionItem definition, DocumentSpan sourceSpan, ReferenceInfoMap referenceInfo)
+        public SourceReferenceItem(DefinitionItem definition, DocumentSpan sourceSpan, ReferenceUsageInfoMap referenceInfo)
         {
             Definition = definition;
             SourceSpan = sourceSpan;
-            ReferenceInfo = referenceInfo ?? ReferenceInfoMap.Empty;
+            ReferenceUsageInfo = referenceInfo ?? ReferenceUsageInfoMap.Empty;
```

Comment:
>ReferenceUsageInfo [](start = 12, length = 18)

We probably need a better name here. This is basically a dictionary of properties, where key is the property name and value is an array of strings. In your case, you have 2 new key-value pairs where keys are property names and values are just single strings instead of array of strings. So both of these are just additional property bags for the SourceReferenceItems with different underlying data structure.

Reply:
Actually, we need to guarantee unique key names across all the additional key-value properties. Maybe the best option would be:

1. Change `s_symbolUsageInfoToReferenceInfoMap` to be `ConcurrentDictionary<SymbolUsageInfo, Immutable<String>>`
2. Store all the additional properties data into the existing `ReferenceInfoMap`. You can wrap the values (i.e. containing type name and member name) into separate single element ImmutableArrays.

---
In reply to: [312664890](https://github.com/dotnet/roslyn/pull/37802#discussion_r312664890) [](ancestors = 312664890)

Reply:
Manish and I discussed offline, 
ReferenceUsageInfo is for properties that can have multiple values (such as read, write, etc...) and can be filtered on those values.
CustomColumns, is for simpler properties that have a single value.  I will change the name to AdditionalProperties.

We recommend a future PR to consolidate these 2 kinds of custom properties into on entity, but in the interest of time would keep it separate for now.  The future PR would guarantee unique key names across the properties.

Reply:
@chborl @mavasani That sounds good to me.  Would it still make sense though to not limit things to "a single value".  i.e. if "AdditionalProperties" were an (immutable) multi-dictionary, then you could basically have arbitary columns (i.e. keys) with any number of values in them per entry. 

Thoughts?

Reply:
@CyrusNajmabadi , that is definitely worth looking into.  I added this to the followup issue, https://github.com/dotnet/roslyn/issues/37961